My last test branch pls

CinyLiang opened the pull request.

gkgeo9 reviewed (commented):
I took a look at your CardGameHintsInteractorTest.java file and wanted to give you some feedback on what I found.

So overall your test structure is pretty solid. Your using proper mocking patterns and testing both success and failure cases which is good. The test method names like successTest and getHintsTest are descriptive but maybe consider being more specific? Like successTest could be testExecuteGeneratesValidHint or somthing so its clearer what your actually testing.

I noticed theres a bunch of commented out code at the bottom starting around the TestCardGameHintsPresenter class. If we're not using it anymore we should probably just remove it to keep things clean.

Your test coverage is actually pretty good. You're testing the success case, various edge cases for extractInner like empty, null, strange input, large input, and exception handling. One thing I noticed though is you dont really test the actual hint generation logic with different card combinations. Like what if we had specific card values, does it generate the right hint format? Maybe something like this would help:

    void testGenerateHintWithKnownCards() {
        TestPresenter presenter = new TestPresenter();
        CardGameHintsInteractor interactor = new CardGameHintsInteractor(presenter);
        List<Card> cards = Arrays.asList(new Card(2), new Card(3), new Card(4), new Card(4));
        
        String hint = interactor.generateHint(cards);
        assertTrue(hint.startsWith("Maybe try"));
        assertTrue(hint.endsWith("first."));
    }

The Clean Architecture stuff looks good. Your properly using the Output Boundary interface, creating Input Data Objects, and keeping the presenter logic seperate which is what we want. One small thing though is in testExceptionWhenGenerateHintFails you're using Mockitos spy which is kind of a code smell in Clean Architecture. The issue is your doing:

    doThrow(new RuntimeException("No solutions found"))
        .when(spyInteractor).generateHint(cards);

This is testing the interactor calling itself which doesnt really test the real failure path. The actual failure would come from SolutionGenerator.find24Solutions() returning an empty list right? Maybe test that scenerio instead of mocking the interactor itself.

Looking at line 15, your TestPresenter has successCalled and failCalled flags which is good, but your just checking assertTrue(successCalled) without really asserting anything about the content of lastOutputData. Like what if the hint is wrong? Should probably verify that too.

The getHintsTest on line 37 just checks that the getter works which is kinda trivial. Not sure if its worth keeping or if you could make it test something more meaningfull.

Your tests dont have any Javadoc comments. While not always required for tests, adding brief descriptions helps especially for the more complex test cases. Something like "Tests that the interactor correctly extracts inner expressions from parenthesized mathematical expressions" would be usefull.

You create cards multiple times in different tests. Might be worth extracting that into a helper method to reduce duplication:

    private CardPuzzle createTestPuzzle(int... values) {
        List<Card> cards = Arrays.stream(values)
            .mapToObj(Card::new)
            .collect(Collectors.toList());
        return new CardPuzzle(cards);
    }

Some suggestions are to add tests for the actual hint quality like does it produce valid mathematical expressions? Test boundary cases for card values like what if all cards are 1 or all are 13? You could also consider using paramaterized tests for the extractInner cases since theyre all testing the same method with different inputs.

Also adding javadoc would be benificial.

Anyways this is solid work overall. The main things would be adding more meaningful assertions, better documentation, and maybe thinking about testing the actual business logic more like does the hint actually help solve the puzzle. Also that spy thing in the exception test should probly be refactored.

CinyLiang commented:
> I took a look at your CardGameHintsInteractorTest.java file and wanted to give you some feedback on what I found.
> 
> So overall your test structure is pretty solid. Your using proper mocking patterns and testing both success and failure cases which is good. The test method names like successTest and getHintsTest are descriptive but maybe consider being more specific? Like successTest could be testExecuteGeneratesValidHint or somthing so its clearer what your actually testing.
> 
> I noticed theres a bunch of commented out code at the bottom starting around the TestCardGameHintsPresenter class. If we're not using it anymore we should probably just remove it to keep things clean.
> 
> Your test coverage is actually pretty good. You're testing the success case, various edge cases for extractInner like empty, null, strange input, large input, and exception handling. One thing I noticed though is you dont really test the actual hint generation logic with different card combinations. Like what if we had specific card values, does it generate the right hint format? Maybe something like this would help:
> 
> ```
> void testGenerateHintWithKnownCards() {
>     TestPresenter presenter = new TestPresenter();
>     CardGameHintsInteractor interactor = new CardGameHintsInteractor(presenter);
>     List<Card> cards = Arrays.asList(new Card(2), new Card(3), new Card(4), new Card(4));
>     
>     String hint = interactor.generateHint(cards);
>     assertTrue(hint.startsWith("Maybe try"));
>     assertTrue(hint.endsWith("first."));
> }
> ```
> 
> The Clean Architecture stuff looks good. Your properly using the Output Boundary interface, creating Input Data Objects, and keeping the presenter logic seperate which is what we want. One small thing though is in testExceptionWhenGenerateHintFails you're using Mockitos spy which is kind of a code smell in Clean Architecture. The issue is your doing:
> 
> ```
> doThrow(new RuntimeException("No solutions found"))
>     .when(spyInteractor).generateHint(cards);
> ```
> 
> This is testing the interactor calling itself which doesnt really test the real failure path. The actual failure would come from SolutionGenerator.find24Solutions() returning an empty list right? Maybe test that scenerio instead of mocking the interactor itself.
> 
> Looking at line 15, your TestPresenter has successCalled and failCalled flags which is good, but your just checking assertTrue(successCalled) without really asserting anything about the content of lastOutputData. Like what if the hint is wrong? Should probably verify that too.
> 
> The getHintsTest on line 37 just checks that the getter works which is kinda trivial. Not sure if its worth keeping or if you could make it test something more meaningfull.
> 
> Your tests dont have any Javadoc comments. While not always required for tests, adding brief descriptions helps especially for the more complex test cases. Something like "Tests that the interactor correctly extracts inner expressions from parenthesized mathematical expressions" would be usefull.
> 
> You create cards multiple times in different tests. Might be worth extracting that into a helper method to reduce duplication:
> 
> ```
> private CardPuzzle createTestPuzzle(int... values) {
>     List<Card> cards = Arrays.stream(values)
>         .mapToObj(Card::new)
>         .collect(Collectors.toList());
>     return new CardPuzzle(cards);
> }
> ```
> 
> Some suggestions are to add tests for the actual hint quality like does it produce valid mathematical expressions? Test boundary cases for card values like what if all cards are 1 or all are 13? You could also consider using paramaterized tests for the extractInner cases since theyre all testing the same method with different inputs.
> 
> Anyways this is solid work overall. The main things would be adding more meaningful assertions, better documentation, and maybe thinking about testing the actual business logic more like does the hint actually help solve the puzzle. Also that spy thing in the exception test should probly be refactored.

hihi Thanks for your comment! Really helpful and detailed feedbacks, I appreciate that!
I have deleted the commented methods and added the two methods you suggested (one testing on a true card set and another testing a true error). I will also add documentations to the tests this weekend.
Regarding the comment about testing whether getHint returns valid hints, it is rather difficult to do so given that this method is based on another method that generates one indeterministic sample solution to the puzzle so what hint returns is also indeterministic. Also, since hint only gives out part of the answer, it is rather hard to validate whether the hint is valid. However, I do have tests for the solution generator in the utilities use case and tests for the hint extraction method extractInner, which I suppose would help validate this getHint method to some extent.
Overall, thanks for your helpful comment!!

wangedison92-sketch reviewed (commented):
Hi, really solid work on this test suite. The coverage here is genuinely impressive — you tested the normal success path, hint generation with specific cards, edge cases like empty or null strings, weird bracket inputs, overly long strings, and even exception behaviour using a Mockito spy. This gives the interactor much more reliability and makes the overall feature feel safer to maintain.

I also like how readable the tests are. Each test method focuses on one idea, and the names are straightforward, which makes it easy to understand at a glance what scenario is being tested. The use of the TestPresenter to track success and fail calls is also a nice touch and keeps the tests decoupled from UI details.

The exception test where you spy on the interactor and force generateHint to fail is especially well done. It shows that you thought about failure paths and didn’t just test “happy cases”, which is something a lot of student projects miss.

One small suggestion (not something that really needs to change)I have is : in the last test where you call execute() with a null puzzle, the test’s logic is a bit long because it manually re-implements part of the interactor’s behaviour. If you ever want to clean it up later, splitting it into two separate tests or letting the interactor fully drive the flow might make it a bit clearer. But again, that’s optional — the current version still works fine and is clear enough.

Overall, this is a very good addition. The tests are thorough, easy to follow, and strengthen the entire card hints feature. Good job!
